src/motioncore/protocols/boolean_gmw/boolean_gmw_gate.cpp

```diff
@@ -764,4 +764,96 @@ const motion::SharePointer MuxGate::GetOutputAsShare() const {
+                                                             Backend& backend)
+    : ConstantAsBooleanGmwInputGate::Base(backend),
+      input_(std::vector(input.begin(), input.end())) {
+  input_owner_id_ = 0;
+  bits_ = input_.size() == 0 ? 0 : input_.at(0).GetSize();
+  InitializationHelper();
+}
+
+ConstantAsBooleanGmwInputGate::ConstantAsBooleanGmwInputGate(std::vector<BitVector<>>&& input,
+                                                             Backend& backend)
+    : ConstantAsBooleanGmwInputGate::Base(backend), input_(std::move(input)) {
+  input_owner_id_ = 0;
+  bits_ = input_.size() == 0 ? 0 : input_.at(0).GetSize();
+  InitializationHelper();
+}
+
+void ConstantAsBooleanGmwInputGate::InitializationHelper() {
+  auto& communication_layer = GetCommunicationLayer();
+  auto& _register = GetRegister();
+
+  if (static_cast<std::size_t>(input_owner_id_) >= communication_layer.GetNumberOfParties()) {
+    throw std::runtime_error(fmt::format("Invalid input owner: {} of {}", input_owner_id_,
+                                         communication_layer.GetNumberOfParties()));
+  }
+
+  gate_id_ = _register.NextGateId();
+
+  assert(input_.size() > 0u);           // assert >=1 wire
+  assert(input_.at(0).GetSize() > 0u);  // assert >=1 SIMD bits
+  // assert SIMD lengths of all wires are equal
+  assert(BitVector<>::IsEqualSizeDimensions(input_));
+
+  if constexpr (kVerboseDebug) {
+    GetLogger().LogTrace(
+        fmt::format("Created a ConstantAsBooleanGmwInputGate with global id {}", gate_id_));
+  }
+
+  output_wires_.reserve(input_.size());
+  for (auto& v : input_) {
+    output_wires_.push_back(GetRegister().EmplaceWire<boolean_gmw::Wire>(v, backend_));
+  }
+
+  if constexpr (kDebug) {
+    auto gate_info = fmt::format("gate id {},", gate_id_);
+    GetLogger().LogDebug(fmt::format(
+        "Created a ConstantAsBooleanGmwInputGate with following properties: {}", gate_info));
+  }
+}
+
+void ConstantAsBooleanGmwInputGate::EvaluateSetup() {}
+
+void ConstantAsBooleanGmwInputGate::EvaluateOnline() {
+  GetBaseProvider().WaitSetup();
+
+  auto& communication_layer = GetCommunicationLayer();
+  auto my_id = communication_layer.GetMyId();
+  auto number_of_parties = communication_layer.GetNumberOfParties();
+
+  std::vector<BitVector<>> result(input_.size());
+  for (auto i = 0ull; i < result.size(); ++i) {
+    if (static_cast<std::size_t>(input_owner_id_) == my_id) {
+      // party 0 holds the constant value
+      result.at(i) = input_.at(i);
+      for (auto party_id = 0u; party_id < number_of_parties; ++party_id) {
+        if (party_id == my_id) {
+          continue;
+        }
+      }
+    } else {
+      // the rest parties hold values of zeros
+      result.at(i) = input_.at(i) ^ input_.at(i);
+    }
+  }
+
+  for (auto i = 0ull; i < output_wires_.size(); ++i) {
+    auto my_wire = std::dynamic_pointer_cast<boolean_gmw::Wire>(output_wires_.at(i));
```

Comment:
wouldn't it be better to have an explicit `Wire` class for constant/plaintext values?

Reply:
ConstantAsBooleanGmwInputGate is used to convert input (publicly known) values into Boolean GMW shares (one party holds the values and the rest parties hold zeros). These Boolean GMW shares are later used as normal Boolean GMW for further computation except that all the parties known these (publicly known) values. In this way, the parties can use (publicly known) values in the computation. 

This is different than the ConstantBooleanInputGate where each party holds exactly the same values.

Reply:
Hmm, then why would you want to have `ConstantAsBooleanGmwInputGate` and not just use the `Party::In` functions?

The argument about SIMD is definitely true. I'm not sure how to solve it w/o breaking the advantages we gain through using SIMD. My feeling is that it needs some more advanced logic to rearrange the wires in a better way, which is probably out of the scope of MOTION, which should really be just an MPC engine that does exactly what you want it to do.

Reply:
According to my understanding, `Party::In` would call input gates (e.g., `proto::boolean_gmw::InputGate`) and that requires generating randomness for secret sharing a value (no matter a private value or constant value). 

But `ConstantAsBooleanGmwInputGate` only requires local operations to "secret share" a constant value $x$ (e.g., party_0 creates an arithmetic gmw share with value $x$, the result parties create an arithmetic gmw share with value $0$). This should bring some efficiency benefits for operations that involves constant values.

Reply:
You can probably reuse `Party::SharedIn(...)` for that.

Reply:
Created function (with tests) `Party::PublicIn` (internal call `Party::In` and `Party::SharedIn`) to input constant values. 
However, for the BMR protocol, I don't find a better way (e.g. only with local operations) than using `Backend::BmrInput`.
Do you have suggestions for sharing constant values in BMR efficiently?

The reason for `ConstantAsBooleanGmwInputGate` rather than `Party::In`:
Suppose a function in ShareWrapper needs to create Boolean GMW shares of constant values. It seems more convenient to call functions (`Backend::ConstantAsBooleanGmwInput` -> `ConstantAsBooleanGmwInputGate`). If use `Party::In`, we need to first get the `PartyPointer` in `ShareWrapper`, which is not very convenient.
Do you think so?